packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,251 @@
+    const segments = normalizedRoute.split('/');
+
+    for (const segment of segments) {
+      // Replace parameterized segments (e.g., :id) with a wildcard (*) for matching
+      const normalizedSegment = segment[0] === ':' ? '*' : segment;
+      let childNode = node.children.get(normalizedSegment);
+
+      if (!childNode) {
+        childNode = this.createEmptyRouteTreeNode(normalizedSegment);
+        node.children.set(normalizedSegment, childNode);
+      }
+
+      node = childNode;
+    }
+
+    // At the leaf node, store the full route and its associated metadata
+    node.metadata = {
+      ...metadata,
+      route: normalizedRoute,
+    };
+
+    node.insertionIndex = this.insertionIndexCounter++;
+  }
+
+  /**
+   * Matches a given route against the route tree and returns the best matching route's metadata.
+   * The best match is determined by the lowest insertion index (earliest defined route).
+   * @param route - The route path to match against the route tree.
+   * @returns The metadata of the best matching route or `undefined` if no match is found.
+   */
+  match(route: string): RouteTreeNodeMetadata | undefined {
+    let bestMatchMetadata: RouteTreeNodeMetadata | undefined;
+    let lowestIndex = Infinity;
+
+    for (const { insertionIndex, metadata } of this.traverseByRoute(route)) {
+      if (insertionIndex < lowestIndex) {
+        bestMatchMetadata = metadata;
+        lowestIndex = insertionIndex;
+      }
+    }
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a serialized format representation.
+   * This method converts the route tree into an array of metadata objects that describe the structure of the tree.
+   * The array represents the routes in a nested manner where each entry includes the route and its associated metadata.
+   *
+   * @returns An array of `RouteTreeNodeMetadata` objects representing the route tree structure.
+   *          Each object includes the `route` and associated metadata of a route.
+   */
+  toObject(): RouteTreeObjectRepresentation {
+    return Array.from(this.traverse());
+  }
+
+  /**
+   * Constructs a `RouteTree` from an object representation.
+   * This method is used to recreate a `RouteTree` instance from an array of metadata objects.
+   * The array should be in the format produced by `toObject`, allowing for the reconstruction of the route tree
+   * with the same routes and metadata.
+   *
+   * @param value - An array of `RouteTreeNodeMetadata` objects that represent the serialized format of the route tree.
+   *                Each object should include a `route` and its associated metadata.
+   * @returns A new `RouteTree` instance constructed from the provided metadata objects.
+   */
+  static fromObject(value: RouteTreeObjectRepresentation): RouteTree {
+    const tree = new RouteTree();
+
+    for (const { route, ...metadata } of value) {
+      tree.insert(route, metadata);
+    }
+
+    return tree;
+  }
```

Comment:
Love how much simpler `traverse` and `traverseByRoute` make these functions. It's so much more intuitive to factor out that iteration logic.